docs/RFCS/raft_consistency_checker.md

```diff
@@ -0,0 +1,114 @@
+used in computing the SHA. The SHA is reported back to the leader through the
+next command.
+2. CollectChecksum: The leader will supply a UUID for the ComputeChecksum
+command and collect the SHA via this command from all replicas. The leader
+computes its own SHA, and waits for a short delay (minutes after sending
+ComputeChecksum), before sending a CollectChecksum, to give the replica plenty
+of time to compute the SHA. If the replica has not computed the SHA it returns
+a NOT_COMPUTED status. In tests this command will block until the checksum is
+computed. If this command never arrives the computed SHA is deleted the next
+time a new ComputeChecksum is seen. 
+
+The periodic consistency checker is run within a scanner that scans through all
+the local replicas and runs the consistency checker for all leader replicas.
+The scanner runs in a continuous loop with equal time spacing between
+replicas such that a single iteration spans the entire periodicity interval. It
+will log an error for all consistency problems seen.
+
+The SHA over a range replica is computed over all the KV pairs using
+replicaDataIterator, ignoring the raft HardState and any log entries with index
+greater than the applied index of the ComputeChecksum command.
+
+Exposing consistency checker to tests/admin:
+
+A cockroach node will support a command through which an admin or a test can
+check the consistency of all ranges for which it is a leader. This will be used
+in all acceptance tests.
+
+Later it will be useful to support a CLI command for an admin to run
+consistency checks over sections of the KV range: e.g., [roachpb.KeyMin,
+roachpb.KeyMax].
+
+Noteworthy scenarios:
+
+1. The leader dies between ComputeChecksum and CollectChecksum: The replicas,
+including the new leader, will continue computing the SHA. The new leader will
+not send a CollectChecksum command.
+2. The leadership change occurs between ComputeChecksum and CollectChecksum:
+Same as 1.
+3. The leader dies and is restored (still as leader) between ComputeChecksum
+and CollectChecksum: The restored leader doesn’t compute the SHA, and the
+replicas never receive CollectChecksum.
+4. The leader dies after sending the CollectChecksum: The new leader might
+replay the CollectChecksum, with each replica reporting its SHA.
+5. A replica dies after receiving ComputeChecksum and receives the
+CollectChecksum later when it comes back up: Since it died it will not have the
+computed SHA, and will reply with a NOT_COMPUTED status.
+6. A replica dies after receiving ComputeChecksum and never receives the
+CollectChecksum: Nothing.
+7. A new replica is added between ComputeChecksum and CollectChecksum: The new
+replica receives the CollectChecksum and returns NOT_COMPUTED.
+8. A replica is removed between ComputeChecksum and CollectChecksum: Nothing.
+9. A range splits or merges with another range after ComputeChecksum: In both
+scenarios the UUID will be invalidated and NOT_COMPUTED will be returned by
```

Comment:
We don't necessarily need to invalidate the checksum on a split; it could be carried forward on the left-hand side to the new range. But this will leave gaps in the sequence; the right-hand side that occurs during a consistency check scan will not be checked unless we have special handling for this. If this is the daily background scan that's probably fine; it'll be picked up in the next cycle. But in a manual check, are these gaps going to be common enough that we need to go back and re-scan ranges that split as we check them? I'm not sure.